src/client_builder.rs

```diff
@@ -0,0 +1,138 @@
+/// ```
+///
+/// Similarly, if using the `rustls-tls` feature you can create a [`Client`] using rustls:
+/// ```no_run
+/// # use imap::ClientBuilder;
+/// # fn main() -> Result<(), imap::Error> {
+/// let client = ClientBuilder::new("imap.example.com", 993).rustls()?;
+/// # Ok(())
+/// # }
+/// ```
+///
+/// To use `STARTTLS`, just call `starttls()` before one of the [`Client`]-yielding
+/// functions:
+/// ```no_run
+/// # use imap::ClientBuilder;
+/// # fn main() -> Result<(), imap::Error> {
+/// let client = ClientBuilder::new("imap.example.com", 993)
+///     .starttls()
+///     .rustls()?;
+/// # Ok(())
+/// # }
+/// ```
+pub struct ClientBuilder<D>
+where
+    D: AsRef<str>,
+{
+    domain: D,
+    port: u16,
+    starttls: bool,
+}
+
+impl<D> ClientBuilder<D>
+where
+    D: AsRef<str>,
+{
+    /// Make a new `ClientBuilder` using the given domain and port.
+    pub fn new(domain: D, port: u16) -> Self {
+        ClientBuilder {
+            domain,
+            port,
+            starttls: false,
+        }
+    }
+
+    /// Use `STARTTLS` for this connection.
+    #[cfg(any(feature = "tls", feature = "rustls-tls"))]
+    pub fn starttls(&mut self) -> &mut Self {
+        self.starttls = true;
+        self
+    }
+
+    /// Return a new [`Client`] using a `native-tls` transport.
+    #[cfg(feature = "tls")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "tls")))]
+    pub fn native_tls(&mut self) -> Result<Client<TlsStream<TcpStream>>> {
+        self.connect(|domain, tcp| {
+            let ssl_conn = TlsConnector::builder().build()?;
+            Ok(TlsConnector::connect(&ssl_conn, domain, tcp)?)
+        })
+    }
+
+    /// Return a new [`Client`] using `rustls` transport.
+    #[cfg(feature = "rustls-tls")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "rustls-tls")))]
+    pub fn rustls(&mut self) -> Result<Client<RustlsStream<TcpStream>>> {
+        self.connect(|domain, tcp| {
+            let ssl_conn = RustlsConnector::new_with_native_certs()?;
+            Ok(ssl_conn.connect(domain, tcp)?)
+        })
+    }
+
+    /// Make a [`Client`] using a custom TLS initialization. This function is intended
+    /// to be used if your TLS setup requires custom work such as adding private CAs
+    /// or other specific TLS parameters.
+    ///
+    /// The `handshake` argument should accept two parameters:
+    ///
+    /// - domain: [`&str`]
+    /// - tcp: [`TcpStream`]
+    ///
+    /// and yield a `Result<C>` where `C` is `Read + Write`. It should only perform
+    /// TLS initialization over the given `tcp` socket and return the encrypted stream
+    /// object, such as a [`native_tls::TlsStream`] or a [`rustls_connector::TlsStream`].
+    ///
+    /// If the caller is using `STARTTLS` and previously called [`starttls`](Self::starttls)
+    /// then the `tcp` socket given to the `handshake` function will be connected and will
+    /// have initiated the `STARTTLS` handshake.
+    ///
+    /// ```no_run
+    /// # use imap::ClientBuilder;
```

Comment:
This probably needs
```rust
#![cfg(feature = "rustls-tls")]
```
So that `cargo test --doc --no-default-features`
```
Doesn't fail.

Reply:
Haha! Almost all the doc tests fail with `cargo test --doc --no-default-features`. 

I fixed all of them. I am not really super happy with all of them? Just using `#![cfg(feature = "xyz")]` to blank out the whole doc test makes the test fail because there is no main and no body. So you have to have something there, so I ended up just scoping the parts of the test that used the features so the test compiles with / without the feature.

The alternative way to do this would be to include some dummy code under `#[cfg(not(feature = "xyz"))]` so it compiles both ways. Do you have a preference?

Reply:
Ugh, that's annoying.. I _feel_ like from memory there's a little bit of magic sauce if you place `#![cfg(feature = "")]` on the very first line of the doctest, but I could be wrong. I do think I'd prefer the `#![cfg]` approach rather than strategically placing extra blocks for `#[cfg()]` though.

Reply:
Yeah, I am kind of surprised that it doesn't just accept an empty doc test block? This seems like the sort of thing that should work, but somehow using `#![cfg(feature = "xyz")]` isn't accepted. I would be happy to be educated here.

Anyway, I pushed something that I think will work and be minimally ugly:

`# {} #[cfg(feature = "tls")]`

The `{}` satisfies the 'something is there' requirement, and the `cfg` line will include / exclude the entire doc `fn main`.

Reply:
That's... awful :sweat_smile: But sure, I still think it's better. I feel like this might be worthwhile to file a bug in rust-lang/rust about actually.

Reply:
Yeah, I find this kind of unsatisfying and it seems like the kind of thing that should work.

I checked through the rust-lang/rust issue tracker but did not see anything like this. If you have messed around with this a bit and confirmed what I am seeing, then we can put together a bug report.

Reply:
I think what I'd do is include in the report:
```rust
/// ```
/// #![cfg(feature = "foo")]
/// assert_eq!(1, 1);
/// ```
pub struct Foo;
```
with the error message suggesting "consider adding a `main` function at the crate level" (which is clearly wrong), and then
```
/// ```rust
/// #![cfg(feature = "foo")]
/// fn main() {
/// assert_eq!(1, 1);
/// }
/// ```
pub struct Foo;
```
which gives the _same_ error even though there _is_ an explicit `fn main` there. At the very least, this is a compiler error problem (cc @estebank), but I would argue it's _also_ a bug/missing feature.

Reply:
@estebank I'm hoping to coax @mordak into doing it since I want more people to get into the habit of reporting unhelpful UX errors, even if they're unsure about whether the underlying feature _should_ be supported. @mordak if you have a chance to write one up, that'd be great, otherwise I'd be happy to do it!

Reply:
No problem, done!

https://github.com/rust-lang/rust/issues/85239